### PR TITLE
Update plot_gate_map() family to leverage graphviz for visualization

### LINKED ISSUE #9031: Update plot_gate_map (and related visualization functions) to leverage rustworkx.visualization.graphviz_draw for unknown coupling maps
### What should we add?

Right now the `plot_gate_map()` visualization function (and similar functions like `plot_circuit_layout()` use rustworkx's `spring_layout()` function when there a backend is passed in and there isn't a hardcoded layout available and passes that to a mpl visualization. For moderate numbers of qubits this works reasonably well, but for large numbers of qubits the layout doesn't scale positions appropriately and the output is typically cluttered and not a useful visualization. Graphviz is particularly well suited for doing this type of graph visualization (as it is a specialized tool for doing graph visualization). So instead of trying to rebuild what we get from graphviz we should just call out to graphviz if it's available and leverage the rustwork function for graph visualization using graphivz to generate the visualization. The only potential issue is potentially `plot_error_map` which has histograms and color bars built using matplotlib and integrating the two different visualization tools might prove difficult.

Here is some example code I used to leverage `graphviz_draw()` to build a view like `plot_circuit_layout()`:

```python
from rustworkx.visualization import graphviz_draw

graph = target.build_coupling_map().graph
for node in graph.node_indices():
    graph[node] = node
    
for edge, triple in graph.edge_index_map().items():
    graph.update_edge_by_index(edge, (triple[0], triple[1]))

physical_bits = layout.get_physical_bits()
qubit_indices = {bit: index for index, bit in enumerate(linear_circuit.qubits)}
    
def color_node(node):
    if node in physical_bits:
        out_dict = {
            "label": str(qubit_indices[physical_bits[node]]),
            "color": "red",
            "fillcolor": "red",
            "style": "filled",
            "shape": "circle",
        }
    else:
        out_dict = {
            "label": "",
            "color": "blue",
            "fillcolor": "blue",
            "style": "filled",
            "shape": "circle",
        }
    return out_dict

def color_edge(edge):
    if all(x in physical_bits for x in edge):
        out_dict = {
            "color": "red",
            "fillcolor": "red",
        }
    else:
        out_dict = {
            "color": "blue",
            "fillcolor": "blue",
        }
    return out_dict
        
graphviz_draw(graph, method="neato", node_attr_fn=color_node, edge_attr_fn=color_edge)
```
which in this case output something like
plot_circuit_layout_graphviz

### PR DESCRIPTION
**Pull request by maxwell04-wq**
### Summary
Fixes #9031 

### Details and comments
- Updated the following modules in `qiskit-terra/qiskit/visualization/gate_map.py` to leverage graphviz plotting for gate maps
    - `plot_gate_map` 
    - `plot_coupling_map`
    - `plot_circuit_layout`
    - `plot_error_map` 
- Uses both `rx` and `matplotlib` for `plot_error_map`, uses only `rx.graphvix_draw` for all other functions.
- ⚠️ The test_gate_map.py file has been updated accordingly. However, the reference images will need to be updated as the tests are failing for maps with larger qubits and no backends (expected behavior as per #9031)
- `qiskit.utils.optionals` does not have a `HAS_RUSTWORKX` check. This can be added to ensure that rustworkx is installed prior to running the `gate_map.py` modules. 
- The `graphviz-draw()` function does not provide the choice to draw directed or undirected graphs. I have left the `plot_directed` parameter in the functions, but as of now it does not affect graph plots. 
- Since `planar_layout()` is not yet available in rustwokx, the graphs have been plotted using `spring_layout()`. This can be modified in future versions.

### CODE REVIEW

**Comment by mtreinish** [comment]
>     * `qiskit.utils.optionals` does not have a `HAS_RUSTWORKX` check. This can be added to ensure that rustworkx is installed prior to running the `gate_map.py` modules.
> 

Rustworkx is a requirement for qiskit, so it will always be installed when using qiskit. You will need a an optional check on `HAS_GRAPHVIZ` though because that will be required to run the function. `rustworkx.graphviz_draw()` has it's own check for this but the exception returned is in a different format so using the qiskit decorator is probably best.

>     * The `graphviz-draw()` function does not provide the choice to draw directed or undirected graphs. I have left the `plot_directed` parameter in the functions, but as of now it does not affect graph plots.

The easiest way to implement this is to use `PyDiGraph.to_undirected()` (https://qiskit.org/documentation/retworkx/apiref/rustworkx.PyDiGraph.to_undirected.html ) to create a `PyGraph` object from a `PyDiGraph`. Then if you pass that to `graphviz_draw()` it'll be an undirected graph visualization.

> 
>     * Since `planar_layout()` is not yet available in rustwokx, the graphs have been plotted using `spring_layout()`. This can be modified in future versions.

TBH, the output of planar layout isn't really suitable for this use case. At the time I added that comment to the code I assumed it was without actually looking (based solely on the name). But this isn't an issue if you're doing a matplotlib based drawing the only real option is to use `spring_layout()` if you're using rustworkx's layout function.

 The thing which might be more interesting to try (and definitely higher quality) instead of rustworkx's `spring_layout` is to leverage graphviz as a layout engine. You can use either the `dot` or `json` output format's from graphviz and then parse that output for the `pos` parameter on each node which you then can feed to the matplotlib drawer (likely after a coordinate system translation).

**Comment by maxwell04-wq** [comment]
@mtreinish I have a question about the tests: I have modified the checks, but need to update the images with which the graphs are compared. Should I wait for a review before changing the figures for the tests?

File: releasenotes/notes/update-gate_map-visualizations-6ea907a0502fdc1a.yaml
Comment (mtreinish) [inline releasenotes/notes/update-gate_map-visualizations-6ea907a0502fdc1a.yaml:11]: One thing I think we need to document here as an `upgrade` note is the API change associated with the return type from the functions modified here. This is potentially a breaking change for any users of these functions because what was previously a `matplotlib.Figure` is now a `PIL.Image` object which behaves very differently. So we need to document this as something that any existing users of `plot_gate_map()` etc will have to change when upgrading to a new version.

Actually just writing that now, I'm wondering if we want to use matplotlib's `imshow` to import the image data into a matplotlib figure and returning that everywhere is the best path here. That would let us continue to return a `Figure` object for backwards compatibility everywhere. What do you think?

**Review by mtreinish — COMMENTED** [review]
This is a great start, thanks for looking at this. I just did a quick high level pass to start and left some comments inline.

File: releasenotes/notes/update-gate_map-visualizations-6ea907a0502fdc1a.yaml
Comment (maxwell04-wq) [reply releasenotes/notes/update-gate_map-visualizations-6ea907a0502fdc1a.yaml:11]: Yes, I agree. Also, since the `plot_error_map()`'s return value is still a matplotlib Figure, this will create inconsistency. I am changing the return types of all the functions back to matplotlib Figure.

File: qiskit/visualization/gate_map.py
Comment (maxwell04-wq) [reply qiskit/visualization/gate_map.py:64]: Changing the return type back to matplotlib figure for all functions
```suggestion
        Figure: A Matplotlib figure instance.
```

File: qiskit/visualization/gate_map.py
Comment (maxwell04-wq) [reply qiskit/visualization/gate_map.py:1132]: No, but it still needs to be a string, as rx only accepts PyStrings for color inputs. Changing to '"#648fff"' for all instances.

File: qiskit/visualization/gate_map.py
Comment (maxwell04-wq) [reply qiskit/visualization/gate_map.py:957]: The plots are being rendered through the `rx.graphviz_draw()` function. I also checked on my system that `graphviz` itself is not necessary to render the plots. Should we revert it back to `HAS_RUSTWORKX`? I am adding this in the `optionals'py` file as well.

File: qiskit/visualization/gate_map.py
Comment (maxwell04-wq) [reply qiskit/visualization/gate_map.py:1132]: Outdated comment. New updates do not need PyStrings for colors.

**Comment by maxwell04-wq** [comment]
@mtreinish thank you for the thorough review of my PR. I have modified my code to reuse the exisiting code. What really helped was using the `mpl_draw` function of `rustworx` instead of `graphviz_draw`. I was also able to define the graph layout using `qubit_coordinates`, and now the rendered figures are much more similar to the original ones, (albeit rotated at a 90º angle). 

One issue that may need to be addressed is using a better function than `spring_layout` for graph layouts. A `grid_layout` function can be developed for rustworkx, much like networkx. Should I open an issue for it? As of now, in the test file, if we increase the similarity tolerance of images in the `test_plot_circuit_layout` function from 0.1 to 0.2 (like the rest of the test functions), the tests will pass. 

Also, as this issue is a part of UnitaryHack, can we try to have the PR merged by June 13? 😄 I realise that it's a close call, but I'll be available at your discretion.

**Comment by mtreinish** [comment]
Out of curiosity I pulled the PR branch locally and ran the script I posted above and this was the output:

`giant_multichip.png`:
giant_multichip_with_pr

`ring.png`:
ring_with_pr

which is better than the layout from the manual mpl code before, but it really shows what I was talking about with the rustworkx `spring_layout` function via `mpl_draw()` not being noticeably different.

**Comment by mtreinish** [comment]
Finally showing the output from graphviz using the basic script I put in #9031:

`giant_multichip.png`:

giant_multichip_graphviz


`ring.png`:

ring_graphviz

**Comment by maxwell04-wq** [comment]
In that case, 

> > @mtreinish thank you for the thorough review of my PR. I have modified my code to reuse the exisiting code. What really helped was using the `mpl_draw` function of `rustworx` instead of `graphviz_draw`. I was also able to define the graph layout using `qubit_coordinates`, and now the rendered figures are much more similar to the original ones, (albeit rotated at a 90º angle).
> > One issue that may need to be addressed is using a better function than `spring_layout` for graph layouts. A `grid_layout` function can be developed for rustworkx, much like networkx. Should I open an issue for it? As of now, in the test file, if we increase the similarity tolerance of images in the `test_plot_circuit_layout` function from 0.1 to 0.2 (like the rest of the test functions), the tests will pass.
> 
> This is the core issue of #9031 that the algorithmic layout functions in rustworkx don't scale for large graphs. This is why I used `graphviz_draw` in the issue as graphviz will be able to layout the graph effectively no matter how large it is. I wasn't too concerned by the format of the visualization changing slightly as a result of using `graphviz_draw`. But if you'd prefer to use `mpl_draw` or alternatively the legacy manual mpl code what I'd recommend is doing what I described the other day is to use graphviz solely as a layout engine ([#10208 (comment)](https://github.com/Qiskit/qiskit-terra/pull/10208#issuecomment-1583152932) ). Then we can leave the rendering to mpl and take advantage of graphviz's better layout.
> 
> My concern with just using `mpl_draw` though is `rx.spring_layout` (or any of the other layout methods available in rustworkx) doesn't scale. We do have an issue to track adding new layout methods: [Qiskit/rustworkx#438](https://github.com/Qiskit/rustworkx/issues/438) so if you're interested we can add a grid layout (I can also help you get started working on the implementation if you're interested in it). But, by itself having a grid layout method in rustworkx isn't really a general solution for #9031 because it assumes a grid like lattice connectivity. For example, if you had a 1000 qubit ring layout a grid layout wouldn't be a good layout.
> 
> I think the best way to test this issue is to build a couple of large qubit count fake backends and test them to see what the graphs look like. For example using `main` if you run:
> 
> ```python
> import numpy as np
> import rustworkx as rx
> 
> from qiskit import transpile, QuantumCircuit
> from qiskit.providers import BackendV2, Options
> from qiskit.transpiler import Target, InstructionProperties
> from qiskit.circuit.library import XGate, SXGate, RZGate, CZGate
> from qiskit.circuit import Measure, Delay, Parameter, IfElseOp
> 
> from qiskit.visualization import plot_gate_map
> 
> class FakeMultiChip(BackendV2):
>     """Fake multi chip backend."""
> 
>     def __init__(self, distance=3, number_of_chips=3):
>         """Instantiate a new fake multi chip backend.
> 
>         Args:
>             distance (int): The heavy hex code distance to use for each chips'
>                 coupling map. This number **must** be odd. The distance relates
>                 to the number of qubits by:
>                 :math:`n = \\frac{5d^2 - 2d - 1}{2}` where :math:`n` is the
>                 number of qubits and :math:`d` is the ``distance``
>             number_of_chips (int): The number of chips to have in the multichip backend
>                 each chip will be a heavy hex graph of ``distance`` code distance.
>         """
>         super().__init__(name='multi_chip')
>         graph = rx.generators.directed_heavy_hex_graph(distance, bidirectional=False)
>         num_qubits = len(graph) * number_of_chips
>         rng = np.random.default_rng(seed=12345678942)
>         rz_props = {}
>         x_props = {}
>         sx_props = {}
>         measure_props = {}
>         delay_props = {}
>         self._target = Target("Fake multi-chip backend", num_qubits=num_qubits)
>         for i in range(num_qubits):
>             qarg = (i,)
>             rz_props[qarg] = InstructionProperties(error=0.0, duration=0.0)
>             x_props[qarg] = InstructionProperties(
>                 error=rng.uniform(1e-6, 1e-4), duration=rng.uniform(1e-8, 9e-7)
>             )
>             sx_props[qarg] = InstructionProperties(
>                 error=rng.uniform(1e-6, 1e-4), duration=rng.uniform(1e-8, 9e-7)
>             )
>             measure_props[qarg] = InstructionProperties(
>                 error=rng.uniform(1e-3, 1e-1), duration=rng.uniform(1e-8, 9e-7)
>             )
>             delay_props[qarg] = None
>         self._target.add_instruction(XGate(), x_props)
>         self._target.add_instruction(SXGate(), sx_props)
>         self._target.add_instruction(RZGate(Parameter("theta")), rz_props)
>         self._target.add_instruction(Measure(), measure_props)
>         self._target.add_instruction(Delay(Parameter("t")), delay_props)
>         cz_props = {}
>         for i in range(number_of_chips):
>             for root_edge in graph.edge_list():
>                 offset = i * len(graph)
>                 edge = (root_edge[0] + offset, root_edge[1] + offset)
>                 cz_props[edge] = InstructionProperties(
>                     error=rng.uniform(1e-5, 5e-3), duration=rng.uniform(1e-8, 9e-7)
>                 )
>         self._target.add_instruction(CZGate(), cz_props)
>         self._target.add_instruction(IfElseOp, name="if_else")
> 
>     @property
>     def target(self):
>         return self._target
> 
>     @property
>     def max_circuits(self):
>         return None
> 
>     @classmethod
>     def _default_options(cls):
>         return Options(shots=1024)
> 
>     def run(self, circuit, **kwargs):
>         raise NotImplementedError("Lasciate ogne speranza, voi ch'intrate")
> 
> 
> class FakeGiantRing(BackendV2):
>     """Fake multi chip backend."""
> 
>     def __init__(self):
>         """Instantiate a new 1k qubit ring backend."""
>         super().__init__(name='multi_chip')
>         graph = rx.generators.directed_cycle_graph(1000)
>         num_qubits = len(graph)
>         rng = np.random.default_rng(seed=12345678942)
>         rz_props = {}
>         x_props = {}
>         sx_props = {}
>         measure_props = {}
>         delay_props = {}
>         self._target = Target("Fake multi-chip backend", num_qubits=num_qubits)
>         for i in range(num_qubits):
>             qarg = (i,)
>             rz_props[qarg] = InstructionProperties(error=0.0, duration=0.0)
>             x_props[qarg] = InstructionProperties(
>                 error=rng.uniform(1e-6, 1e-4), duration=rng.uniform(1e-8, 9e-7)
>             )
>             sx_props[qarg] = InstructionProperties(
>                 error=rng.uniform(1e-6, 1e-4), duration=rng.uniform(1e-8, 9e-7)
>             )
>             measure_props[qarg] = InstructionProperties(
>                 error=rng.uniform(1e-3, 1e-1), duration=rng.uniform(1e-8, 9e-7)
>             )
>             delay_props[qarg] = None
>         self._target.add_instruction(XGate(), x_props)
>         self._target.add_instruction(SXGate(), sx_props)
>         self._target.add_instruction(RZGate(Parameter("theta")), rz_props)
>         self._target.add_instruction(Measure(), measure_props)
>         self._target.add_instruction(Delay(Parameter("t")), delay_props)
>         cz_props = {}
>         for root_edge in graph.edge_list():
>             edge = (root_edge[0], root_edge[1])
>             cz_props[edge] = InstructionProperties(
>                 error=rng.uniform(1e-5, 5e-3), duration=rng.uniform(1e-8, 9e-7)
>             )
>         self._target.add_instruction(CZGate(), cz_props)
>         self._target.add_instruction(IfElseOp, name="if_else")
> 
>     @property
>     def target(self):
>         return self._target
> 
>     @property
>     def max_circuits(self):
>         return None
> 
>     @classmethod
>     def _default_options(cls):
>         return Options(shots=1024)
> 
>     def run(self, circuit, **kwargs):
>         raise NotImplementedError("Lasciate ogne speranza, voi ch'intrate")
> 
> 
> if __name__ == "__main__":
>     backend_multichip = FakeMultiChip(15, 5)
>     backend_ring = FakeGiantRing()
>     plot_gate_map(backend_multichip).savefig('giant_multichip.png')
>     plot_gate_map(backend_ring).savefig('ring.png')
> ```
> 
> this is what you get out today:
> 
> `giant_multichip.png`: giant_multichip
> 
> `ring.png`: ring
> 
> which is less than useful :) While this output is with the current `main` branch, I expect this PR's branch using `mpl_draw()` to have similar output because it's relying on the same layout function under the covers. That script might be helpful in testing any potential solution to see how the proposed change performs on larger scale backends.

In that case, should I drop the `qubit_coordinates` parameter? If yes, the only remaining changes will be converting the PIL images of `ghraphviz_draw()` to matplotlib images and use the previous color schemes. Also, in that case, the color hex values will need to be reconverted to PyStrings.

**Comment by mtreinish** [comment]
> @mtreinish I have reverted the functions to use `graphviz_draw()`. The return type of all functions is same as the original one, i.e., a matplotlib figure unless `ax` is specified. However, one major change is the absence of `qubit_coordinates` as they are not needed anymore to render the plots.
> 
> Currently, plenty of tests are breaking because I have removed the `qubit_coordinates` parameter from the `plot_coupling_map` function. Please let me know how you'd like it to be modified.

We should be able to keep the `qubit_coordinates` argument functional. With `neato` there is a `pos` attribute you can set on a node to provide a position: https://graphviz.org/docs/attrs/pos/ also when paired to `pin` (https://graphviz.org/docs/attrs/pin/) it prevents neato from moving a node during it's layout. So to keep `qubit_coordinates` all we need to do is convert the input format to the function to something that graphviz expects (which likely also involves mapping the coordinates to different origin points) and setting that in the node attr function when `qubit_coordinates` is set.

**Comment by maxwell04-wq** [comment]
@mtreinish I have incorporated the `pos` parameter in the plots, and the functions can now use the `qubit_coordinates` parameter as well. Please review the code; in the meantime, I'll try to fix the broken tests.

File: test/python/quantum_info/operators/symplectic/test_clifford.py
Comment (mtreinish) [inline test/python/quantum_info/operators/symplectic/test_clifford.py:589]: This looks like an unintentional debugging change.

Comment (maxwell04-wq) [reply test/python/quantum_info/operators/symplectic/test_clifford.py:589]: Yes, I'll remove it.

File: qiskit/visualization/gate_map.py
Comment (mtreinish) [inline qiskit/visualization/gate_map.py:23]: These imports will need to be done at run time, because matplotlib isn't guaranteed to be avaliable when this module is imported. So we'll need these imports inside the functions where they're used to avoid a potential import error for users that don't have matplotlib installed.

I actually tested doing this locally in a windows vm with python 3.8 and it also fixed the segfault errors that windows CI is failing on. @jakelishman and I were discussing this offline and think it is likely a side effect of a stack limit being hit by having extra modules in the import path somehow (the exact path isn't clear). But anyway moving these imports to be done at runtime in the functions should fix the CI failures too.

Comment (mtreinish) [reply qiskit/visualization/gate_map.py]: I went ahead and fixed this in: https://github.com/Qiskit/qiskit-terra/pull/10208/commits/d719620000dc1d3e4114fa876828865a7485b7b5

File: qiskit/visualization/gate_map.py
Comment (mtreinish) [inline qiskit/visualization/gate_map.py:1084]: I feel like this isn't needed here, since it's done in the `if qubit_coordinates` case above, but it also doesn't hurt since `pin` doesn't have any meaning unless we set `pos` according to https://graphviz.org/docs/attrs/pin/

```suggestion
```

File: releasenotes/notes/update-gate-map-visualizations-6ea907a0502fdc1a.yaml
Comment (mtreinish) [inline releasenotes/notes/update-gate-map-visualizations-6ea907a0502fdc1a.yaml:7]:
```suggestion
    The visualizations from the :func:`~.plot_gate_map`,  :func:`~.plot_coupling_map`.
    :func:`~.plot_error_map`,  and :func:`~.plot_circuit_layout` functions have been significantly
    improved for rendering layouts of backends with large numbers of qubits. This was accomplished
    by leveraging `graphviz <https://graphviz.org/>`__ through rustworkx's ``graphviz_draw()`` function
    to perform a more sophisticated algorithmic graph layout that scales for large numbers of
    qubits.
```

File: releasenotes/notes/update-gate-map-visualizations-6ea907a0502fdc1a.yaml
Comment (mtreinish) [inline releasenotes/notes/update-gate-map-visualizations-6ea907a0502fdc1a.yaml:19]:
```suggestion
    The visualization functions: :func:`~.plot_gate_map`,  :func:`~.plot_coupling_map`.
    :func:`~.plot_error_map`,  and :func:`~.plot_circuit_layout` now depend on
    `graphviz <https://graphviz.org/>`__ being installed to function. This change was
    necessary to enable visualizing backends with larger numbers of qubits. This
    additional external requirement is in addition to the existing optional dependencies
    these functions previously required. You find details on how to install
    graphviz here:  https://graphviz.org/download/
```

**Review by mtreinish — APPROVED** [review]
Overall the code LGTM now, I think this is ready to merge. just a couple small inline comments updating the release notes and removing one stray argument which shouldn't be needed anymore.

**Comment by mtreinish** [comment]
Hmm, it looks like there are some CI updates still needed here because #9961 fixed the image comparison tests so we need to update the reference images for the new output this causes. Also it looks like the windows segfault is still tripping us up here, although the runtime import fixed this locally for. Locally I moved one extra import when it fixed the segfault, but I didn't do that for the PR because I didn't think it was really needed. I'll try that here and hopefully that'll fix it for good.

File: qiskit/visualization/gate_map.py
Comment (mtreinish) [inline qiskit/visualization/gate_map.py:1048]: The image comparison tests caught this edge case, we need to flatten parallel edges here with:

```suggestion
        graph = graph.to_undirected(multigraph=False)
```

see: https://qiskit.org/documentation/retworkx/apiref/rustworkx.PyDiGraph.to_undirected.html

Otherwise graphs that should look like:

65_qubit_gate_map

end up looking like:

65_qubit_gate_map

**Comment by mtreinish** [comment]
The other bug that the image comparison tests (which are now failing after #9961) caught is a 1 qubit coupling map. Previously it would be visualized as: 

1_qubit_gate_map

but now it looks like an empty image file:

1_qubit_gate_map

(it's there it's just a small white square)

**Comment by jakelishman** [comment]
I've made #10415 to track the issue that's causing the segfaults on Windows.  The reason it relates to this PR is that this PR adds some `optionals.HAS_SEABORN` checks, and importing `seaborn` causes `jedi` to be imported, which sets the recursion limit higher than Windows can actually support.  This PR is doing the right thing and shouldn't need to change.

**Comment by mtreinish** [comment]
I fixed the bug with the 1 qubit backend drawing (the `CouplingMap(coupling).graph` method to get the rustworkx graph wouldn't work because there are no edges in a 1 qubit graph). But I think I'm going to remove this from the 0.25 milestone in the interest of time since we're tagging 0.25.0rc1 tomorrow. The only thing really left is the image comparison tests, most of those will just need an update since the images aren't going to be identical anymore. But there are some weird things going on with the font scaling in some cases. For example, with this PR one of the visual tests is generating 

16_qubit_gate_map

and the reference image is:

16_qubit_gate_map

besides being mirrored and different edges and node sizes (all of which I think are fine) the issue is the font is getting clipped by the node's circle. I think we should fix this before merging.

**Comment by maxwell04-wq** [comment]
> 

Sure. Looking into it.

**Comment by maxwell04-wq** [comment]
> I fixed the bug with the 1 qubit backend drawing (the `CouplingMap(coupling).graph` method to get the rustworkx graph wouldn't work because there are no edges in a 1 qubit graph). But I think I'm going to remove this from the 0.25 milestone in the interest of time since we're tagging 0.25.0rc1 tomorrow. The only thing really left is the image comparison tests, most of those will just need an update since the images aren't going to be identical anymore. But there are some weird things going on with the font scaling in some cases. For example, with this PR one of the visual tests is generating
> 
> 16_qubit_gate_map
> 
> and the reference image is:
> 
> 16_qubit_gate_map
> 
> besides being mirrored and different edges and node sizes (all of which I think are fine) the issue is the font is getting clipped by the node's circle. I think we should fix this before merging.

I've tested the new font size on a test code. Does this work?

image

**Review by mtreinish — APPROVED** [review]
I just made some small tweaks to the formatting and fixed the font size issues. Also I updated the reference images in the visual comparison tests. I think this is ready to go now. Thanks a whole lot for doing this and sticking with it through all the updates. This project turned out to be a bit more finicky than it first seemed.

I think we'll probably be dealing with small formatting details for a while after making this change, but the gains from using a proper graph visualization engine are going to be a net win.

**Pull request merged**
